src/jsonrpc-virtual-machine.cpp

```diff
@@ -48,90 +57,153 @@ std::string jsonrpc_post_data(const std::string &method, const std::tuple<Ts...>
+    if (!stream.socket().is_open()) {
+        // Connect
+        while (true) {
+            beast::error_code ec;
+            stream.connect(remote_endpoint, ec);
+            if (!ec) { // Success
+                break;
+            } else if (ec == asio::error::interrupted) {
+                // Retry the operation during interrupts (SIGINT/SIGTERM),
+                // otherwise we may leave dead zombies processes during fork requests.
+            } else { // Unexpected error
+                throw beast::system_error(ec);
+            }
+        }
+
+        // Disable Nagle's algorithm to minimize TCP connection latency
+        const boost::asio::ip::tcp::no_delay no_delay_option(true);
+        stream.socket().set_option(no_delay_option);
+
+        // Minimize socket close time by setting the linger time to 0.
+        // On MacOS, it avoids accumulating socket in TIME_WAIT state, after rapid successive requests,
+        // which can consume all available ports.
+        // It is safe to do this because it is the client who decides to close the connection,
+        // after all data is received.
+        const boost::asio::socket_base::linger linger_option(true, 0);
+        stream.socket().set_option(linger_option);
+
+        // Enable keep alive TCP option for keep alive HTTP connection
+        if (keep_alive) {
+            const boost::asio::socket_base::keep_alive keep_alive_option(true);
+            stream.socket().set_option(keep_alive_option);
+        }
     }
-    if (data.status_code != "200") {
-        throw std::runtime_error("http error: "s + data.reason_phrase + " (code "s + data.status_code + ")"s);
+
+    try {
+        // Set up a HTTP request message
+        http::request<http::string_body> req;
+        req.method(http::verb::post); // POST
+        req.version(11);              // Only HTTP 1.1 support keep alive connections
+        req.target(uri);
+        req.keep_alive(keep_alive);
+        req.set(http::field::host, host);
+        req.set(http::field::content_type, "application/json");
+        req.set(http::field::user_agent, BOOST_BEAST_VERSION_STRING);
+        req.body() = post_data;
+        req.prepare_payload();
+
+        // Send the HTTP request
+        while (true) {
+            beast::error_code ec;
+            http::write(stream, req, ec);
+            if (!ec) { // Success
+                break;
+            } else if (ec == asio::error::interrupted) {
+                // Retry the operation during interrupts (SIGINT/SIGTERM),
+                // otherwise we may leave dead zombies processes during fork requests.
+            } else { // Unexpected error
+                throw beast::system_error(ec);
+            }
+        }
+
+        // Set up HTTP response parser
+        beast::flat_buffer buffer;
+        http::response_parser<http::string_body> res_parser;
+        res_parser.eager(true);
+        res_parser.body_limit(16777216U); // can receive up to 16MB
```

Comment:
Does this mean we will fail with a "read_memory" call for a larger chunk?

Reply:
Yes. The same was true with Mongoose before, where we defined this:
```
# The biggest message is reset uarch with large_data: 2x 4MB + small data + overhad
MONGOOSE_DEFS+=-DMG_MAX_RECV_SIZE=12582912
```
So in our Mongoose implementation we could receive up to 12MB.

Boost Beast by default limits to 8MB, so this code duplicates the maximum, tests would fail with 8MB. If the message is bigger than that, the request will fail.

Reply:
Should we limit the local API as well to 8M? Or do we allow them to differ in behavior?

Reply:
I don't think we should limit the local API for reading/writing more than 16MB, this is something I personally do for local machines (for testing and hacking purposes). The remote machine is mostly used by the node, and I don't think the node will ever need to read more than 16MB, so I personally don't mind having them different for this specific case. If someone complains we could raise the limit in the future.

Reply:
I am worried about the send_cmio_response call. Who knows when somebody will want to dehash a large chunk of data? We should keep this in mind and advertise the limitation. Not sure either if the node will use jsonrpc or not.

Reply:
What kind of error message does the client get when trying to, say, write_memory(20MB)?